### ISSUE TITLE
Allow to send local time(zone)

### ISSUE THREAD

**Issue by aaannndddyyy** [irungentoo/toxcore#1431]
A useful feature that tox is still missing is to show the local time of a contact, if he opted for disclosing it.
(Should be stored and visible also when the contact is offline)

**Comment by x368** [comment]
-1 for an addition to the protocol

Actually this is a limited case of tracking the location of the contact. The matter is complicated, involving both privacy and access control (disclosure by contact and by occasion, sometimes you will want to lie and possibly different versions to different contacts), even when this is "just" a time zone.

I'd rather see it like an UI issue.

It would serve all practical purposes if the owner of the contact list (and no one else) could indicate the time zones of the contacts, which would allow indicating their corresponding local time _according to the idea of the owner of the contact list_.

As much as you want to know and your contacts agree to disclose or update this information they simply can tell you. This is both convenient - no extra menus _they_ have to visit - and flexible, allows to disclose or not (with opt-in rather than opt-out), change at will and of course allows to tell different things to different contacts.

**Comment by aaannndddyyy** [comment]
skype has it, tox strives to be a slype replacement. If you have many contacts you possibly don't want to ask all of them.
If a user does not want to disclose his whereabouts, he can simply disable it.
a per-user setting would overcomplicate things, imho.
If he really only wants to allow a select set of users to know it, he disables it as he wants to hide it, and if those ask he replies.

**Comment by x368** [comment]
> skype has it, tox strives to be a slype replacement.

There is no point in replacing if we are to imitate everything, including the design flaws. Skype does _not_ care about privacy, quite the opposite.

> If you have many contacts you possibly don't want to ask all of them.

Then just do not ask all of them, ask only those you need this info about.

It is _your_ desire, not theirs, why do you want to make _them_ bother whom they allow to see the timezone, and whom not?

> a per-user setting would overcomplicate things, imho

You are on the way to oversimplify, which can be even more harmful.

This _will_ lead the majority of the users into believing that "allow or disallow for all" is the only way to do this.

Such oversimplification is certainly most harmful if it lands in the core protocol. Some clients might choose such a "simplified" _representation_ anyway - imho bad by itself but not as bad, if there will be other clients to compare to.

Let us look at the related research:

'When someone tells you, "People don't care about privacy," consider whether the game has been designed and rigged so that they cannot care about privacy' -  Alessandro Acquisti, Carnegie-Mellon University, while presenting a research about how easily humans are mislead in privacy-related issues. You can see the whole presentation (15 min) at http://www.ted.com/talks/alessandro_acquisti_why_privacy_matters.

In other words, it is _really_ important to avoid "simple" reasoning like

"I want this person to know my time, thus let me open to anyone"

Even a prominent reminder "ALLOWING THIS CAN HARM YOU LATER!" will _not_ do - the user will forget about it... how soon? let you see the TED report above, really!

(from the same talk: "_any_ personal information can become sensitive information.")

TL;DR  users should __not__ be offered potentially insecure behavior as default, humans are really (_really_) bad at estimating consequences of sharing their information.

> This could actually even be expanded to general profile info, like place too.

Unless every piece of information in the profile info has its own access list (and choosing "accessible to everyone" is made hard) this will be misleading.

Then this will potentially lead to unsolvable situation when the user wants to continue allowing the access for her contacts, but for some reason wants some of them to see _different_ information.

Generalizing / Offtopic:

If we really want to make provisions for sharing "quite a lot" of information (well, why not?), this should be robust and hence adjustable per contact. Note that switching profiles/identities is not always applicable.

I would suggest maintaining a "main" version of one's metainformation (sex, age, marital status, place, timezone, preferred language, religion, disabilities, underwear color...) and exposing it to contacts _only_ by an explicit action and so that one can create multiple versions of the same information piece and give different contacts references to different versions of each field. This sounds complicated but it _is_ complicated even though Skype and others may willingly oversimplify (they do not risk anything, the users do).

I do not say that we _have_ to implement this but we should have a good conceptual model before we put more support for information sharing into the core protocol.

**Comment by aaannndddyyy** [comment]
Are you also proposing getting rid of avatar support then?
People might put in photos. And I cannot set the avatar on a per-contact basis…
I honestly think you are overdramatizing. 
If you don't use tor or vpn you can be geolocated anyway. 
This info I am suggesting here is only sent privately to people in your contactlist - just like your nick, your status message (which could reveal even your mood and not just time), avatar. 
And I think it is save to assume that people who know how to set up tor or a vpn can have a look at their tox settings and disable it. It should be in the privacy section, so you cannot miss it.
so 
1) It is a nice-to-have feature for average user
2) It is normally not very sensitive information
3) It is sent only to your friends, in an encrypted way
4) You can decide whether to send it or not - so nobody is forced to do it.

While I see your point that you might want some to know and others not, I don't think it's a big issue.
Most of those I wouldn't want it to see are different kind of friends. I mean they must not know where I live, so I would also not want them to see me online whenever my normal friends see me online, so I would simply use a different profile. 
As soon as multi-profile support is there, this will be pretty easy.
I would not mind a per-profile setting. But it would be tedious to set it one by one.

I don't think underwear is a needed piece of information.
Maybe place is not either. When I wrote place I somehow had skype's search unconsciously in mind. There if you are looking for Tom Brown, and there are several, you'd be glad to know the place, but you are right, that outside a user search, this does not make very much sense.
But I have friends spread all over the globe and I keep forgetting how much is the time offset for each of them - especially with daylight saving times etc.
(I don't wanna look it up over and over agasin, ask them over and over again, or maintan a spreadsheet for that)

I don't see why you would want different versions of the info.
I don't think lies are a good think to foster. If you don't wanna tell your timezone, just don't tell it, but don't spread lies.

**Comment by x368** [comment]
Thanks for taking the time to answer in detail.

> And I cannot set the avatar on a per-contact basis…

This is an arbitrary limitation inherited from somewhere else, isn't it?

> If you don't use tor or vpn you can be geolocated anyway. 

This is actually irrelevant. I may wish to tell some of my contacts where I am even when my computer is hidden behind proxies and vice versa I may not care whether someone tries to geolocate me but not show my time zone to certain people who do not care to geolocate.

And finally if you geolocated my IP this is your _guess_ about my location. If I told you about my time zone this is a _statement_. The difference is quite large.

> This info ... just like your nick, your status message (which could reveal even your mood and not just time), avatar.

Exactly! It is actually pretty dumb that everyone sees the same mood (or avatar) of mine. I can be at the same time "flirty" for my girlfriend and "tired" for my buddy who brags about being sooo drunk :)

Same for other properties, even the time zone.

> 1) It is a nice-to-have...

Yes!

> 2) It is normally not very sensitive information

Yes but _normally_ should not mean that _when it is sensitive_ it would hinder the "whole" functionality (for other contacts of the user).

> 3) It is sent only to your friends

I assume that you will not insist on this statement :) Are the bank clerk / your boss / your customer - always your friends?

> 4) You can decide whether to send it or not

According to your proposal - only to "send it to all" or "never to anybody" which makes the user quite powerless. If one wants her lover to know her mood^H^H^H^Htimezone but not her other "friend" who would certainly spread the news about her travelling... what to do??

> But I have friends spread all over the globe and I keep forgetting how much is the time offset for each of them - especially with daylight saving times etc

No problem to let a Tox client remember their time zones and _visualize_ their time!

Nothing says _they_ have to _publish_ the zones. Set a time zone once per friend is hardly a remarkable burden (for the person who wants to be reminded about it).

> I don't think underwear is a needed piece of information.

You do not - but how do we know what is valuable for others? Any choice is arbitrary, who is to decide? The core has an important responsibility to _not_ introduce arbitrary limitations.

> I don't think lies are a good think to foster. If you don't wanna tell your timezone, just don't tell it, but don't spread lies.

It is easy to be misled by emotionally charged words like a "lie". I do not have as good a neutral substitution as in "friend"->"contact". Let's see by example. May be you do not want to tell your parents that you are traveling overseas - to avoid them to be afraid for imaginary dangers lurking for you. But if you suddenly disallow them to see your time zone, what happens? Will they be calm? This is just one example, the mood above is another. There are (lots of) others of course.

**Comment by x368** [comment]
sorry missed to address your point about multi-profile support.

It mitigates the problem as a workaround but does not solve it.

As one example, I may want to talk at the same time with two other persons (group chat)  and still present to them different information about my mood, underwear, time zone, avatar...

That's why I wrote
> Note that switching profiles/identities is not always applicable.

**Comment by GrayHatter** [comment]
@irungentoo close as dup, request to send meta data
